northstar/src/runtime/mod.rs

```diff
@@ -272,42 +272,68 @@ async fn runtime_task(config: &'_ Config, stop: CancellationToken) -> Result<(),
         event_tx.send(Event::Shutdown).await.ok();
     });
 
+    let mut shutdown_issued = false;
+
     // Enter main loop
     loop {
-        if let Err(e) = match event_rx.recv().await.unwrap() {
+        match event_rx.recv().await {
+            // If a shutdown event was issued, any remaning client request will be responded
+            // with an error message
+            Some(Event::Console(_msg, txr)) if shutdown_issued => {
+                txr.send(api::model::Response::Err(api::model::Error::Console(
+                    "Shutting down".to_string(),
+                )))
+                .ok();
+            }
             // Debug console commands are handled via the main loop in order to get access
             // to the global state. Therefore the console server receives a tx handle to the
             // main loop and issues `Event::Console`. Processing of the command takes place
             // in the console module but with access to `state`.
-            Event::Console(msg, txr) => state.console_request(&msg, txr).await,
+            Some(Event::Console(msg, txr)) => state.console_request(&msg, txr).await?,
             // The OOM event is signaled by the cgroup memory monitor if configured in a manifest.
             // If a out of memory condition occours this is signaled with `Event::Oom` which
             // carries the id of the container that is oom.
-            Event::Oom(container) => state.on_oom(&container).await,
+            Some(Event::Oom(container)) => state.on_oom(&container).await?,
             // A container process existed. Check `process::wait_exit` for details.
-            Event::Exit(container, exit_status) => state.on_exit(&container, &exit_status).await,
+            Some(Event::Exit(container, exit_status)) => {
+                state.on_exit(&container, &exit_status).await?;
+            }
             // The runtime os commanded to shut down and exit.
-            Event::Shutdown => break state.shutdown().await,
+            Some(Event::Shutdown) => {
+                shutdown_issued = true;
+
+                // TODO ask what we want to do here
+                // debug!("Stop accepting new connections");
+                // if let Some(console) = &console {
+                //     console.stop_new_connctions();
+                // }
+
+                debug!("Shutting down runtime");
+                state.shutdown().await.unwrap();
+
+                debug!("Closing event loop");
+                event_rx.close();
```

Comment:
Closing `event_rx` will make this loop break. The flag `shutdown_issued` is never read with value `true`.

Reply:
https://docs.rs/tokio/1.5.0/tokio/sync/mpsc/struct.Receiver.html#method.close The documentation mentions that a call to `close` will not drop the receiver, i.e. all the events generated in `state.shutdown()` will remain there (e.g. all the stop containers events and notifications). The `shutdown_issued` flag is used in the first arm of this match expression to reply to all those possibly queued console requests with an error.